### PR TITLE
SceneSpawner: add unload() and unload_sync()

### PR DESCRIPTION
**Pull request by Fishrock123**
This allows you to unload scenes.

Refs: https://github.com/bevyengine/bevy/issues/255

I've been struggling to get an example up that isn't just rewriting the breakout game to use scenes, maybe someone can help me make something simpler to properly verify this.

### CODE REVIEW

File: crates/bevy_scene/src/scene_spawner.rs [hunk 1]
```diff
@@ -152,19 +179,32 @@ impl SceneSpawner {
...
             }
         }
 
-        self.scenes_to_load = non_existent_scenes;
+        Ok(())
+    }
+
+    pub fn unload_queued_scenes(&mut self, world: &mut World) -> Result<(), SceneSpawnError> {
+        let mut scenes_to_unload = Vec::new();
+        std::mem::swap(&mut self.scenes_to_unload, &mut scenes_to_unload);
```
Comment (memoryruins) [inline crates/bevy_scene/src/scene_spawner.rs:200]:
```suggestion
        let scenes_to_unload = std::mem::take(&mut self.scenes_to_unload);
```

File: crates/bevy_scene/src/scene_spawner.rs [hunk 2]
```diff
@@ -152,19 +179,32 @@ impl SceneSpawner {
...
+
+        for scene_handle in scenes_to_unload {
+            match self.unload_sync(world, scene_handle) {
+                Ok(_) => {}
+                Err(err) => return Err(err),
+            }
```
Comment (memoryruins) [inline crates/bevy_scene/src/scene_spawner.rs:206]:
```suggestion
            self.unload_sync(world, scene_handle)?;
```

File: crates/bevy_scene/src/scene_spawner.rs [hunk 3]
```diff
@@ -173,19 +213,19 @@ impl SceneSpawner {
         world: &mut World,
         resources: &Resources,
     ) -> Result<(), SceneSpawnError> {
-        let scenes_to_spawn = self.scenes_to_spawn.drain(..).collect::<Vec<_>>();
-        let mut non_existent_scenes = Vec::new();
+        let mut scenes_to_spawn = Vec::new();
+        std::mem::swap(&mut self.scenes_to_spawn, &mut scenes_to_spawn);
```
Comment (memoryruins) [inline crates/bevy_scene/src/scene_spawner.rs:217]:
```suggestion
        let scenes_to_spawn = std::mem::take(&mut self.scenes_to_spawn);
```

File: crates/bevy_scene/src/scene_spawner.rs (hunk 1 above)
Comment (Fishrock123) [reply crates/bevy_scene/src/scene_spawner.rs:200]: Ah, that's handy. Good to know.

File: crates/bevy_scene/src/scene_spawner.rs [hunk 4]
```diff
@@ -58,6 +63,28 @@ impl SceneSpawner {
...
+        if let Some(instance_ids) = self.spawned_scenes.get(&scene_handle) {
+            for instance_id in instance_ids {
+                if let Some(instance) = self.spawned_instances.get(&instance_id) {
+                    for entity in instance.entity_map.values() {
+                        let _ = world.despawn(*entity); // Ignore the result, unload only cares if it exists.
+                    }
+                }
+            }
+
+            self.loaded_scenes
```
Comment (cart) [inline crates/bevy_scene/src/scene_spawner.rs:80]: lets replace these retain() calls with
```rust
self.loaded_scenes.remove(&scene_handle);
```

These are hashsets/hashmaps, which means we don't need to iterate over the whole collection to find a key.

File: crates/bevy_scene/src/scene_spawner.rs [hunk 5]
```diff
@@ -152,19 +179,27 @@ impl SceneSpawner {
...
-        let scenes_to_load = self.scenes_to_load.drain(..).collect::<Vec<_>>();
-        let mut non_existent_scenes = Vec::new();
+        let scenes_to_load = std::mem::take(&mut self.scenes_to_load);
+
         for scene_handle in scenes_to_load {
             match self.load_sync(world, resources, scene_handle) {
                 Ok(_) => {}
                 Err(SceneSpawnError::NonExistentScene { .. }) => {
-                    non_existent_scenes.push(scene_handle)
+                    self.scenes_to_spawn.push(scene_handle)
```
Comment (cart) [inline crates/bevy_scene/src/scene_spawner.rs:188]: this should be `self.scenes_to_load`

**Review by cart — CHANGES_REQUESTED** [review]
Looks good. A lot of people want this feature :smile: 

Just a couple of small changes needed. I'd say its fine to not add an example in this pr. Its worth spending some time to come up with an interesting spawn / despawn example.

**Comment by Fishrock123** [comment]
@cart Addressed and updated.

**Review by cart — APPROVED** [review]

**Pull request merged**